pkg/agent/run.go

```diff
@@ -235,16 +238,30 @@ func gatherData(ctx context.Context, config Config) []*api.DataReading {
 					len(es), strings.Join(points, "\n\t"))
 			}
 		}
-		return dgError.ErrorOrNil()
+		return dgError
 	}
 
-	err := backoff.RetryNotify(getReadings, backOff, notify)
-	if err != nil {
-		log.Println(err)
-		log.Printf("This will not be retried")
+	if StrictMode {
+		multiError := getReadings()
+		// check all the errors and find if it's due to a missing dg or not
+		for _, err := range multiError.WrappedErrors() {
+			if !strings.Contains(err.Error(), "the server could not find the requested resource") {
```

Comment:
This part of the code should not worry about specific text on the errors.

[The data-gatherer](https://github.com/jetstack/preflight/blob/c62dbc4b54642565ceff374d8cc9826980d703e6/pkg/datagatherer/k8s/dynamic.go#L138) should return an appropriate error type.

Also, in general, we should not be checking the text in error messages, as that is fragile. First I would check if there is an error type we can check for in the upstream k8s client. If not, the more robust solution would be to check if the CRD is defined or not. If that turns out to be too complicated, then checking the text of the error might be a good choice. But again, that should be a responsibility of the data-gatherer, not `agent/run.go`.

Reply:
Yes, that is what I meant. Ultimately, the decision of "failing hard vs retrying" will be made on `agent/run.go`. But I would not expect that code to compare the error text. Ideally, it would just check an error type.